### DISCUSSION TITLE
TextBox Pattern Highlighting

### DISCUSSION THREAD

**Discussion by stevemonaco** [AvaloniaUI/Avalonia#17077] [Q&A] [answered]
** Edit: I've updated the sample to also include a simple implementation of `TextBox` highlighting. **

I recently released a demo for [`TextBlock` highlighting](https://github.com/stevemonaco/AvaloniaDemos/tree/master/TextHighlighting). I want to extend the demo to include highlighting for `TextBox`, but ran into some issues. The core rendering logic fits cleanly within a derived `TextBlock` or `SelectableTextBlock`:

<details>

<summary>Rendering</summary>

```cs
protected void RenderHighlightRanges(DrawingContext context, Point origin)
{
	if (_highlightRangeCache is null || Highlight is null)
		return;

	foreach (var range in _highlightRangeCache)
	{
		var rects = TextLayout.HitTestTextRange(range.Index, range.Length);

		using (context.PushTransform(Matrix.CreateTranslation(origin)))
		{
			foreach (var rect in rects)
			{
				context.FillRectangle(Highlight, PixelRect.FromRect(rect, 1).ToRect(1));
			}
		}
	}
}

protected override void RenderTextLayout(DrawingContext context, Point origin)
{
	RenderHighlightRanges(context, origin);
	base.RenderTextLayout(context, origin);
}
```

</details>

However, this is not the case for `TextBox` because it renders text via `TextPresenter` which leads to the following problems:

1. `TextPresenter.Render` is `sealed` and `TextPresenter.RenderInternal` is `private`. Therefore I cannot derive from `TextPresenter` to add render logic.
2. `TextBox` requires a `TextPresenter` to operate. If I recreate `TextPresenter` as a new type, it would fail even after re-templating.

**I'm looking for clues on how to implement this or the scope of an acceptable PR.** My current preferred app-side implementation route would be: 1. derive from `TextPresenter` to add rendering logic. 2. re-template `TextBox` so that my derived `TextPresenter` is used.

Exposing some pre/post rendering callbacks would work, too, but aren't really in the style of Avalonia. I could also copy/paste `TextBox` as a new control, but that requires about half a dozen other internal components, too.

**Comment by hamster620** [comment, marked as answer]
You may refer to [SyntaxHighlightingTextBox.cs](https://github.com/carina-studio/AppSuiteBase/blob/master/SyntaxHighlighting/Controls/SyntaxHighlightingTextBox.cs) which is currently used by [ULogViewer](https://github.com/carina-studio/ULogViewer).
![image](https://github.com/user-attachments/assets/7cc83900-a1f4-4c7b-ae0d-16c4441959b8)

**Reply by stevemonaco** [reply]
Ah, there is enough surface area for a custom presenter if you override with replacement instead of extension and create the text layout. I'll have to check if there's enough to render the highlight myself or if I'm stuck with inlines. Both projects look quite nice. Thanks!
